<hr><h1 align='center'>Project: Segmenting and Clustering Neighborhoods in Toronto</h1>

<h1 align='center'>IBM Data Science Professional Certificate</h1><br><hr>

<h3 align='center'> Author: Sebastiano Fazzino</h3><hr>

<br><h4 align='center'>This notebook is part of 'Applied Data Science Capstone' course. In this notebook I'll be exploring and clustering the neighborhoods in Toronto.

We start installing <strong>'geopy'</strong> and <strong>'lxml'</strong> trhough <strong>'!pip install'</strong>

In [13]:
!pip install geopy
!pip install lxml==4.5.2


Then we import all the libraries required for the the project:

In [14]:
import pandas as pd
import numpy as np
from geopy.geocoders import Nominatim
from sklearn.cluster import KMeans
import folium
import requests 
from pandas.io.json import json_normalize 
import matplotlib.cm as cm
import matplotlib.colors as colors
%matplotlib inline

<br>In this project we'll be working on a database stored on a Wikipedia page, so we'll need to scrape the page, wrangle the data and read it into a pandas dataframe. Luckily <strong>Pandas</strong> can do all the job for us in just two rows of code.<br>
Analyzing the page https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M we can notice that the database we need is stored into the first table in the page, so we proceed as follows:

In [15]:
#We store the page into 'table' variable
table = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
#We store just the first table of the page (table[0]) into 'toronto_df'
toronto_df = table[0]

In [16]:
#We print the first five rows of our dataframe
toronto_df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

<br>As we can see there are some 'Not Assigned' values in 'Borough' and 'Neighbourhood' columns, so we use <strong>'group_by'</strong> & <strong>'count'</strong> functions to check how many 'Not Assigned' values are there in total in our dataframe

In [17]:
print(toronto_df.groupby('Borough').count())
print(toronto_df.shape)

                  Postal Code  Neighbourhood
Borough                                     
Central Toronto             9              9
Downtown Toronto           19             19
East Toronto                5              5
East York                   5              5
Etobicoke                  12             12
Mississauga                 1              1
North York                 24             24
Not assigned               77             77
Scarborough                17             17
West Toronto                6              6
York                        5              5
(180, 3)


<br>We have a total of 77 rows with 'Not Assigned' values that we need to drop:

In [18]:
df_toronto = toronto_df[toronto_df['Borough'] != 'Not assigned'].reset_index(drop=True)
df_toronto

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

We can easily verify that we dropped the unwanted rows as the shape of the dataframe went down from 180 rows to 103 rows.

<br><br>In the next step we want to know the coordinates for each neighborhood according to its 'Postal Code'. We have all the coordinates stored in a csv file that we can read into a pd dataframe using <strong>'pd.read_csv'</strong> function:

In [19]:
coordinates = pd.read_csv('Geospatial_Coordinates.csv')
coordinates

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

We can now combine the two dataframes in a new one:

In [20]:
df_new = pd.merge(df_toronto,coordinates, on = "Postal Code")
df_new

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

<br><br>Using <strong>Geocoder</strong> we can find the coordinates of Toronto

In [21]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [22]:
Toronto_Map = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, lng, borough, neighborhood in zip(df_new['Latitude'], df_new['Longitude'], df_new['Borough'], df_new['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='yellow',
        fill=True,
        fill_color='red',
        fill_opacity=0.7,
        parse_html=False).add_to(Toronto_Map)  
    
Toronto_Map


<br><br>In the next step we'll define <strong>Foursquare</strong> credentials and version

In [23]:
CLIENT_ID = 'YDEFE3BRIVVDPTQPTR042NEQONEDSTGFRVSAH0QNOAOAUQF4'
CLIENT_SECRET = '0K5OYCFOP0OMPFTJCSPEPOXQ2NSUSMAP0SRMOZWBYZQRBTEN'
VERSION = '20200803'

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: YDEFE3BRIVVDPTQPTR042NEQONEDSTGFRVSAH0QNOAOAUQF4
CLIENT_SECRET:0K5OYCFOP0OMPFTJCSPEPOXQ2NSUSMAP0SRMOZWBYZQRBTEN


<br><br>Using <strong>Foursquare</strong> we create a function to find the top 100 venues for each neighborhood in Toronto

In [24]:
lat = df_new['Latitude']
lng = df_new['Longitude']
radius = 500
limit = 100

def getTorontoVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        toronto_url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            limit)
            
        results = requests.get(toronto_url).json()["response"]['groups'][0]['items']
        
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [25]:
toronto_venues = getTorontoVenues(names=df_new['Neighbourhood'],
                                   latitudes=df_new['Latitude'],
                                   longitudes=df_new['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

<br><br>Let's check how many venues were returned for each neighborhood:

In [26]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            4   
Alderwood, Long Branch                                               9   
Bathurst Manor, Wilson Heights, Downsview North                     19   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   24   
...                                                                ...   
Willowdale, Willowdale East                                         34   
Willowdale, Willowdale West                                          6   
Woburn                                                               4   
Woodbine Heights                                                     6   
York Mills West                                                      4   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             4   
Alderwood, Long Branch                                                9   
Bathurst Manor, Wilson Heights, Downsview North                      19   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    24   
...                                                                 ...   
Willowdale, Willowdale East                                          34   
Willowdale, Willowdale West                                           6   
Woburn                                                                4   
Woodbine Heights                                                      6   
York Mills West                                                       4   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            4               4   
Alderwood, Long Branch                               9               9   
Bathurst Manor, Wilson Heights, Downsview North     19              19   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   24              24   
...                                                ...             ...   
Willowdale, Willowdale East                         34              34   
Willowdale, Willowdale West                          6               6   
Woburn                                               4               4   
Woodbine Heights                                     6               6   
York Mills West                                      4               4   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      4   
Alderwood, Long Branch                                         9   
Bathurst Manor, Wilson Heights, Downsview North               19   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             24   
...                                                          ...   
Willowdale, Willowdale East                                   34   
Willowdale, Willowdale West                                    6   
Woburn                                                         4   
Woodbine Heights                                               6   
York Mills West                                                4   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     4  
Alderwood, Long Branch                                        9 

<br><br>We use <strong>'get_dummies'</strong> method to convert categorical variables into dummy variables and we store the information in a new dataframe 


In [27]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]


<br><br>Let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [28]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Agincourt          0.0   
1                            Alderwood, Long Branch          0.0   
2   Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                   Bayview Village          0.0   
4                 Bedford Park, Lawrence Manor East          0.0   
..                                              ...          ...   
90                      Willowdale, Willowdale East          0.0   
91                      Willowdale, Willowdale West          0.0   
92                                           Woburn          0.0   
93                                 Woodbine Heights          0.0   
94                                  York Mills West          0.0   

    Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                 0.0                0.0      0.0                 0.0   
1                 0.0                0.0      0.0                 0.0   
2                 0.0                0.0      0.0                 0.0   
3                 0.0                0.0      0.0                 0.0   
4                 0.0                0.0      0.0                 0.0   
..                ...                ...      ...                 ...   
90                0.0                0.0      0.0                 0.0   
91                0.0                0.0      0.0                 0.0   
92                0.0                0.0      0.0                 0.0   
93                0.0                0.0      0.0                 0.0   
94                0.0                0.0      0.0                 0.0   

    Airport Gate  Airport Lounge  Airport Service  Airport Terminal  ...  \
0            0.0             0.0              0.0               0.0  ...   
1            0.0             0.0              0.0               0.0  ...   
2            0.0             0.0              0.0               0.0  ...   
3            0.0             0.0              0.0               0.0  ...   
4            0.0             0.0              0.0               0.0  ...   
..           ...             ...              ...               ...  ...   
90           0.0             0.0              0.0               0.0  ...   
91           0.0             0.0              0.0               0.0  ...   
92           0.0             0.0              0.0               0.0  ...   
93           0.0             0.0              0.0               0.0  ...   
94           0.0             0.0              0.0               0.0  ...   

    Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0     0.0            0.0                            0.0               0.0   
1     0.0            0.0                            0.0               0.0   
2     0.0            0.0                            0.0               0.0   
3     0.0            0.0                            0.0               0.0   
4     0.0            0.0                            0.0               0.0   
..    ...            ...                            ...               ...   
90    0.0            0.0                            0.0               0.0   
91    0.0            0.0                            0.0               0.0   
92    0.0            0.0                            0.0               0.0   
93    0.0            0.0                            0.0               0.0   
94    0.0            0.0                            0.0               0.0   

    Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0      0.000000               0.000000              0.0       0.0   
1      0.000000               0.000000              0.0       0.0   
2      0.000000               0.000000              0.0       0.0   
3      0.000000               0.000000              0.0       0.0   
4      0.000000               0.000000              0.0       0.0   
..          ...                    ...              ...       ...   
90     0.000000         

<br><br>Let's print each neighborhood along with the top 10 most common venues

In [29]:
num_top_venues = 10

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0                     Lounge  0.25
1               Skating Rink  0.25
2  Latin American Restaurant  0.25
3             Breakfast Spot  0.25
4             Medical Center  0.00
5             Massage Studio  0.00
6          Martial Arts Dojo  0.00
7   Mediterranean Restaurant  0.00
8        Monument / Landmark  0.00
9                Men's Store  0.00


----Alderwood, Long Branch----
                       venue  freq
0                Pizza Place  0.22
1                        Gym  0.11
2               Skating Rink  0.11
3                   Pharmacy  0.11
4                Coffee Shop  0.11
5             Sandwich Place  0.11
6                        Pub  0.11
7                       Pool  0.11
8                Yoga Studio  0.00
9  Middle Eastern Restaurant  0.00


----Bathurst Manor, Wilson Heights, Downsview North----
                 venue  freq
0                 Bank  0.11
1          Coffee Shop  0.11
2       Ice Cream Shop  0.05
3    

In [30]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

<br><br>Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [31]:
top_10_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighborhood']
for ind in np.arange(top_10_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head(10)

Neighborhood  1st Most Common Venue  \
0                                          Agincourt                 Lounge   
1                             Alderwood, Long Branch            Pizza Place   
2    Bathurst Manor, Wilson Heights, Downsview North                   Bank   
3                                    Bayview Village     Chinese Restaurant   
4                  Bedford Park, Lawrence Manor East        Thai Restaurant   
5                                        Berczy Park            Coffee Shop   
6                        Birch Cliff, Cliffside West  General Entertainment   
7       Brockton, Parkdale Village, Exhibition Place                   Café   
8  Business reply mail Processing Centre, South C...     Light Rail Station   
9  CN Tower, King and Spadina, Railway Lands, Har...        Airport Service   

       2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant               Skating Rink        Breakfast Spot   
1                   Pharmacy                        Pub                  Pool   
2                Coffee Shop  Middle Eastern Restaurant      Sushi Restaurant   
3                       Café                       Bank   Japanese Restaurant   
4             Sandwich Place                Coffee Shop    Italian Restaurant   
5         Seafood Restaurant                Cheese Shop          Cocktail Bar   
6            College Stadium               Skating Rink                  Café   
7                Coffee Shop                     Bakery        Breakfast Spot   
8              Auto Workshop                       Park            Comic Shop   
9             Airport Lounge           Airport Terminal           Coffee Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0               Dog Run          Dessert Shop    Dim Sum Restaurant   
1        Sandwich Place           Coffee Shop          Skating Rink   
2        Ice Cream Shop         Shopping Mall     Mobile Phone Shop   
3         Women's Store          Dessert Shop    Dim Sum Restaurant   
4            Restaurant              Pharmacy     Indian Restaurant   
5                Bakery            Restaurant              Beer Bar   
6   Distribution Center      Department Store          Dessert Shop   
7          Climbing Gym          Intersection               Stadium   
8           Pizza Place      Recording Studio            Restaurant   
9         Boat or Ferry       Harbor / Marina               Airport   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Diner        Discount Store    Distribution Center  
1                   Gym   Distribution Center         Discount Store  
2         Deli / Bodega            Restaurant    Fried Chicken Joint  
3                 Diner        Discount Store    Distribution Center  
4                   Pub               Butcher                   Café  
5                  Café        Farmers Market          Shopping Mall  
6    Dim Sum Restaurant                 Diner         Discount Store  
7    Italian Restaurant            Restaurant          Burrito Place  
8         Burrito Place               Brewery             Skate Park  
9    Airport Food Court          Airport Gate               Boutique

<br><br>We can now cluster the neighborhoods in our dataframe. We use <strong>'k-mean'</strong> to cluster the neighborhoods into 4 clusters


In [45]:
kcluster = 5
k_means = KMeans(init = "k-means++", n_clusters = kcluster, n_init = 8)

In [46]:
toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)
k_means.fit(toronto_grouped_clustering)

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
    n_clusters=5, n_init=8, n_jobs=None, precompute_distances='auto',
    random_state=None, tol=0.0001, verbose=0)

In [47]:
k_means_labels = k_means.labels_
k_means_labels

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4,
       0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 2, 0, 0, 4, 0,
       0, 0, 3, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 3, 0, 0, 0, 3,
       0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 3,
       0, 3, 0, 0, 4, 0, 3], dtype=int32)

<br><br>To conclude, we visualize the resulting clusters:

In [48]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

x = np.arange(kcluster)
ys = [i + x + (i*x)**2 for i in range(kcluster)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(df_new['Latitude'], df_new['Longitude'], df_new['Neighbourhood'], k_means_labels):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.5).add_to(map_clusters)
       
map_clusters